libraries/lib-music-information-retrieval/MirDsp.cpp

```diff
@@ -0,0 +1,159 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  MirDsp.cpp
+
+  Matthieu Hodgkinson
+
+**********************************************************************/
+#include "MirDsp.h"
+#include "FFT.h"
+#include "IteratorX.h"
+#include "MirAudioReader.h"
+#include "MirTypes.h"
+#include "MirUtils.h"
+#include "StftFrameProvider.h"
+
+#include <cassert>
+#include <cmath>
+#include <numeric>
+
+namespace MIR
+{
+namespace
+{
+float GetNoveltyMeasure(
+   const std::vector<float>& prevPowSpec, const std::vector<float>& powSpec)
+{
+   auto k = 0;
+   return std::accumulate(
+      powSpec.begin(), powSpec.end(), 0., [&](float a, float mag) {
+         // Half-wave-rectified stuff
+         return a + std::max(0.f, mag - prevPowSpec[k++]);
+      });
+}
+
+std::vector<float> GetMovingAverage(const std::vector<float>& x, double hopRate)
+{
+   constexpr auto smoothingWindowDuration = 0.2;
+   // An odd number.
+   const int M = std::round(smoothingWindowDuration * hopRate / 4) * 2 + 1;
+   const auto window = GetNormalizedHann(2 * M + 1);
+   auto n = 0;
+   std::vector<float> movingAverage(x.size());
+   std::transform(x.begin(), x.end(), movingAverage.begin(), [&](float) {
+      const auto m = IotaRange(-M, M + 1);
+      const auto y =
+         std::accumulate(m.begin(), m.end(), 0., [&](double y, int i) {
+            auto k = n + i;
+            if (k < 0)
+               k += x.size();
+            else if (k >= x.size())
+               k -= x.size();
+            return y + x[k] * window[i + M];
+         });
+      ++n;
+      // The moving average of the raw ODF will be subtracted from it to yield
+      // the final ODF, negative results being set to 0. (This is to remove
+      // noise of small ODF peaks before the method's quantization step.) The
+      // larger this multiplier, the less peaks will remain. This value was
+      // found by trial and error, using the benchmarking framework
+      // (see TatumQuantizationFitBenchmarking.cpp)
+      constexpr auto thresholdRaiser = 1.5;
```

Comment:
I've done bottom-up review of all these calculations, twice now.  I'm doing one more pass, top-down, depth first.

Now I think it would be clearer if moved this multiplication by 1.5, into MirDsp where the moving average is used.

Then this so-called `GetMovingAverage` will really do what the function name suggests, and only that.